cmd/lotus-shed/postfind.go

```diff
@@ -0,0 +1,123 @@
+	lapi "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	lcli "github.com/filecoin-project/lotus/cli"
+	"github.com/filecoin-project/specs-actors/v2/actors/builtin"
+	"github.com/urfave/cli/v2"
+)
+
+var postFindCmd = &cli.Command{
+	Name:        "post-find",
+	Description: "return addresses of all miners who have over zero power and have posted in the last day",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "tipset",
+			Usage: "specify tipset state to search on",
+		},
+		&cli.BoolFlag{
+			Name:  "verbose",
+			Usage: "get more frequent print updates",
+		},
+		&cli.BoolFlag{
+			Name:  "withpower",
+			Usage: "only print addrs of miners with more than zero power",
+		},
+	},
+	Action: func(c *cli.Context) error {
+		api, acloser, err := lcli.GetFullNodeAPI(c)
+		if err != nil {
+			return err
+		}
+		defer acloser()
+		ctx := lcli.ReqContext(c)
+		verbose := c.Bool("verbose")
+		withpower := c.Bool("withpower")
+
+		ts, err := lcli.LoadTipSet(ctx, c, api)
+		if err != nil {
+			return err
+		}
+		if ts == nil {
+			ts, err = api.ChainHead(ctx)
+			if err != nil {
+				return err
+			}
+		}
+		oneDayAgo := ts.Height() - abi.ChainEpoch(2880)
+		if verbose {
+			fmt.Printf("Collecting messages between %d and %d\n", ts.Height(), oneDayAgo)
+		}
+		// Get all messages over the last day
+		msgs := make([]*types.Message, 0)
+		for ts.Height() > oneDayAgo {
+			// Get messages on ts parent
+			next, err := api.ChainGetParentMessages(ctx, ts.Cids()[0])
+			if err != nil {
+				return err
+			}
+			msgs = append(msgs, messagesFromAPIMessages(next)...)
+
+			// Next ts
+			ts, err = api.ChainGetTipSet(ctx, ts.Parents())
+			if err != nil {
+				return err
+			}
+			if verbose && int64(ts.Height())%100 == 0 {
+				fmt.Printf("Collected messages back to height %d\n", ts.Height())
+			}
+		}
+		fmt.Printf("Loaded messages to height %d\n", ts.Height())
+
+		mAddrs, err := api.StateListMiners(ctx, ts.Key())
+		if err != nil {
+			return err
+		}
+
+		minersToCheck := make(map[address.Address]struct{})
+		for _, mAddr := range mAddrs {
+			// if they have no power ignore. This filters out 14k inactive miners
+			// so we can do 100x fewer expensive message queries
+			if withpower {
+				power, err := api.StateMinerPower(ctx, mAddr, ts.Key())
+				if err != nil {
+					return err
+				}
+				if power.MinerPower.RawBytePower.GreaterThan(big.Zero()) {
+					minersToCheck[mAddr] = struct{}{}
+				}
+			} else {
+				minersToCheck[mAddr] = struct{}{}
+			}
+		}
+		fmt.Printf("Loaded %d miners with power\n", len(minersToCheck))
+
+		postedMiners := make(map[address.Address]struct{})
+		for _, msg := range msgs {
+			_, hasPower := minersToCheck[msg.To]
+			_, seenBefore := postedMiners[msg.To]
+
+			if hasPower && !seenBefore {
+				if msg.Method == builtin.MethodsMiner.SubmitWindowedPoSt {
+					fmt.Printf("%s\n", msg.To)
+					minersToCheck[msg.To] = struct{}{}
+				}
+			}
```

Comment:
This needs to check who the message is being sent to as actors can reuse method numbers, though probably better to do this else where when collection messages?

Reply:
I think that's covered by `hasPower := minersToCheck[msg.To]`, which means it must be a miner. And the addresses come from `api.StateListMiners`.

Reply:
Do you mean the type of actor we're sending to?  We know it must be a miner because we've seen msg.To in the minersToCheck map

Reply:
Sounds good!